### PR TITLE
Enable GVN for `AggregateKind::RawPtr`

### LINKED ISSUE #124792: MIR const-prop not putting `usize` into `AggregateKind::RawPtr`?
Discord Context: https://discord.com/channels/273534239310479360/273541522815713281/1236936970912137236

I was surprised in <https://rust.godbolt.org/z/hnMT6K4qh> to see
```rust
        _18 = const 128_usize;
        StorageLive(_24);
        _24 = _22 as *const () (PtrToPtr);
        _23 = *const [u8] from (_24, _18);
```

Because I would have expected constprop into the Aggregate, like
```rust
        StorageLive(_24);
        _24 = _22 as *const () (PtrToPtr);
        _23 = *const [u8] from (_24, const 128_usize);
```

As happened in other places, like
```rust
        _20 = Offset(_16, const 128_usize);
        StorageLive(_21);
        _21 = Sub(_13, const 128_usize);
```

`AggregateKind::RawPtr` is relatively new (#123840), so maybe it needs to be handled properly somewhere else?

### PR DESCRIPTION
**Pull request by scottmcm**
Looks like I was worried for nothing; this seems like it's much easier than I was originally thinking it would be.
r? @cjgillot 

This should be useful for `x[..4]`-like things, should those start inlining enough to expose the lengths.

### CODE REVIEW

File: compiler/rustc_mir_transform/src/gvn.rs [hunk 1]
```diff
@@ -176,6 +176,9 @@ enum AggregateTy<'tcx> {
     Array,
     Tuple,
     Def(DefId, ty::GenericArgsRef<'tcx>),
+    /// The type of the pointer can't be determined from the operands,
+    /// so we have to keep the full thing here.
+    RawPtr(Ty<'tcx>),
```
Comment (WaffleLapkin) [inline compiler/rustc_mir_transform/src/gvn.rs:181]: it's this the type of the pointer, or the pointee? (`*const A` vs `A`)

File: compiler/rustc_mir_transform/src/gvn.rs [hunk 2]
```diff
@@ -399,7 +403,13 @@ impl<'body, 'tcx> VnState<'body, 'tcx> {
                     };
                     for (field_index, op) in fields.into_iter().enumerate() {
                         let field_dest = self.ecx.project_field(&variant_dest, field_index).ok()?;
-                        self.ecx.copy_op(op, &field_dest).ok()?;
+                        if let AggregateTy::RawPtr { .. } = kind && field_index == 0 {
+                            // The pointer field might be any thin pointer type,
+                            // so avoid ICEing for a type mismatch.
+                            self.ecx.copy_op_allow_transmute(op, &field_dest).ok()?;
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/gvn.rs:409]: This code path has no test. Could you add one?

**Review by cjgillot — COMMENTED** [review]
Thanks. For the desconstruction operations (cast and Len for now), do you plan to do it as a follow-up, or do you want me to do it?

File: compiler/rustc_mir_transform/src/gvn.rs [hunk 3]
```diff
@@ -960,6 +1007,9 @@ impl<'body, 'tcx> VnState<'body, 'tcx> {
             (UnOp::Not, Value::BinaryOp(BinOp::Ne, lhs, rhs)) => {
                 Value::BinaryOp(BinOp::Eq, *lhs, *rhs)
             }
+            (UnOp::PtrMetadata, Value::Aggregate(AggregateTy::RawPtr { .. }, _, fields)) => {
+                return Some(fields[1]);
+            }
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/gvn.rs:1012]: Should we add such a line to `Rvalue::Len` too?

File: compiler/rustc_mir_transform/src/gvn.rs [hunk 4]
```diff
@@ -920,15 +937,45 @@ impl<'body, 'tcx> VnState<'body, 'tcx> {
...
+            {
+                fields[0] = *cast_value;
+                *data_pointer_ty = *cast_from;
+                was_updated = true;
+            }
+
+            if was_updated && let Some(local) = self.try_as_local(fields[0], location) {
+                field_ops[FieldIdx::ZERO] = Operand::Copy(Place::from(local));
+                self.reused_locals.insert(local);
+            }
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/gvn.rs:977]: I usually check both constants and locals, to prefer keeping a constant when one is available. Or is it too unlikely here?

**Comment by cjgillot** [comment]
Thanks!
@bors r+

**Pull request merged**